src/web/app/components/dialogs/promark/ZAxisAdjustment.tsx

```diff
@@ -0,0 +1,239 @@
+
+interface Props {
+  device: IDeviceInfo;
+  onClose: () => void;
+}
+
+export const ZAxisAdjustment = ({ device, onClose }: Props): JSX.Element => {
+  const { global: tGlobal, promark_settings: t } = useI18n();
+  const { model } = device;
+  const isInch = useMemo(() => storage.get('default-units') === 'inches', []);
+  const [zAxis, setZAxis] = useState(1);
+  const [isPreviewing, setIsPreviewing] = useState(false);
+  const [isMoving, setIsMoving] = useState(false);
+  const [parameters, setParameters] = useState<MarkParameters>({ power: 100, speed: 350 });
+  const previewTask = useRef<string>('');
+  const markTask = useRef<string>('');
+  const waitTillNotRunning = async () => {
+    await deviceMaster.waitTillStatusPredicate({
+      predicate: (status) => status !== deviceConstants.status.RUNNING,
+    });
+  };
+  const movingTimeout = useRef<NodeJS.Timeout>();
+
+  const uploadPreviewTask = useCallback(async () => {
+    if (!previewTask.current) {
+      previewTask.current = await generateCalibrationTaskString({ width: 10 });
+    }
+
+    await loadTaskToSwiftray(previewTask.current, model);
+  }, [model]);
+
+  useEffect(() => {
+    markTask.current = '';
+  }, [parameters]);
+
+  const uploadMarkTask = useCallback(async () => {
+    const { power, speed } = parameters;
+
+    if (!markTask.current) {
+      markTask.current = await generateCalibrationTaskString({ width: 10, power, speed });
+    }
+
+    await loadTaskToSwiftray(markTask.current, model);
+  }, [model, parameters]);
+
+  useEffect(() => {
+    const abortPreview = async () => {
+      setIsPreviewing(false);
+    };
+
+    swiftrayClient.on('disconnected', abortPreview);
+
+    return () => {
+      swiftrayClient.off('disconnected', abortPreview);
+    };
+  }, []);
+
+  const handleMove = async (distance: number) => {
+    const getTime = (distance: number) => (Math.abs(distance) * 1000) / 3 + 500;
+
+    if (isPreviewing) {
+      await deviceMaster.stopFraming();
+      await waitTillNotRunning();
+      setIsPreviewing(false);
+    }
+
+    setIsMoving(true);
+
+    await deviceMaster.adjustZAxis(distance);
+
+    movingTimeout.current = setTimeout(() => {
+      setIsMoving(false);
+    }, getTime(distance));
+  };
+
+  const handlePreview = async () => {
+    if (!isPreviewing) {
+      await uploadPreviewTask();
+      await deviceMaster.startFraming();
+      setIsPreviewing(true);
+    } else {
+      await deviceMaster.stopFraming();
+      await waitTillNotRunning();
+      setIsPreviewing(false);
+    }
+  };
+
+  const handleMark = async () => {
+    if (isPreviewing) {
+      await deviceMaster.stopFraming();
+      await waitTillNotRunning();
+      setIsPreviewing(false);
+    }
+
+    await uploadMarkTask();
+    await deviceMaster.doPromarkCalibration();
+  };
+
+  const handleStop = () => {
+    const stop = async () => {
+      try {
+        if (isPreviewing) {
+          await deviceMaster.stopFraming();
```

Comment:
Do we need add
```
await waitTillNotRunning();
setIsPreviewing(false);
```
here as well like other places?

Reply:
yes, the stop(Framing) is not execute immediately.
for example, if user stop preview and then mark , 
the rabbit would jump out instantly.